File: lib/features/order/services/order_service.dart
import 'package:kiosk_flutter/config.dart';
import 'package:kiosk_flutter/models/backend_exceptions.dart';
import 'package:kiosk_flutter/models/backend_models.dart';
import 'package:kiosk_flutter/models/menu_item_with_description.dart';
import 'package:kiosk_flutter/utils/supabase/aps_description_repository.dart';
import 'package:kiosk_flutter/utils/supabase/aps_order_repository.dart';
import 'package:kiosk_flutter/utils/supabase/item_description_repository.dart';
import 'package:kiosk_flutter/utils/supabase/menu_repository.dart';
import 'package:kiosk_flutter/utils/supabase/supabase_function_repository.dart';
import 'package:rxdart/rxdart.dart';

class OrderService {
  OrderService({
    required ApsOrderRepository orderRepository,
    required MenuRepository menuRepository,
    required ItemDescriptionRepository itemRepository,
    required ApsDescriptionRepository apsRepository,
    required SupabaseFunctionRepository supabaseFunctionRepository,
  })  : _orderRepository = orderRepository,
        _menuRepository = menuRepository,
        _itemRepository = itemRepository,
        _apsRepository = apsRepository,
        _supabaseFunctionRepository = supabaseFunctionRepository;

  final ApsOrderRepository _orderRepository;
  final ItemDescriptionRepository _itemRepository;
  final MenuRepository _menuRepository;
  final ApsDescriptionRepository _apsRepository;
  final SupabaseFunctionRepository _supabaseFunctionRepository;

  final BehaviorSubject<ApsOrder?> _orderSubject = BehaviorSubject<ApsOrder?>();

  Future<void> createNewOrder() async {
    final newOrder = ApsOrder(
      apsId: AppConfig.instance.apsId,
      origin: OriginType.kiosk,
      status: OrderStatus.duringOrdering,
      estimatedTime: 0,
      kdsOrderNumber: null,
      pickupNumber: null,
      createdAt: DateTime.now(),
      updatedAt: DateTime.now(),
    );

    final createdOrder = await _orderRepository.createApsOrder(order: newOrder);
    _orderSubject.add(createdOrder);
  }

  // Stream zamówienia z repozytorium, który nasłuchuje na zmiany w zamówieniu
  Stream<ApsOrder?> get orderStream {
    return _orderSubject.switchMap((order) {
      if (order != null && order.id != null) {
        return _orderRepository.streamApsOrder(orderId: order.id!);
      } else {
        return Stream.value(null);
      }
    });
  }

  // Stream dla elementów zamówienia
  Stream<List<ApsOrderItem>> apsOrderItemsStream() {
    return _orderSubject.switchMap((order) {
      if (order != null && order.id != null) {
        return _orderRepository.streamApsOrderItems(orderId: order.id!);
      } else {
        return Stream.value(<ApsOrderItem>[]);
      }
    });
  }

  // Zamknij stream, kiedy nie jest potrzebny
  void dispose() {
    _orderSubject.close();
  }

  Stream<List<AvailableItem>> availableItemStream({
    required int apsId,
  }) {
    final apsStream = _apsRepository.apsStream(apsId: apsId);

    return apsStream.switchMap((apsDescription) {
      final menuId = apsDescription.menuId;

      final menuItemPricesStream = _menuRepository.menuItemPriceStream(menuId: menuId);
      final apsOrdersStream = _orderRepository.streamAllApsOrders();
      final apsOrderItemsStream = _orderRepository.streamAllApsOrderItems();

      // Łączenie trzech strumieni w jeden (menuItems, APS Orders i APS Order Items)
      return Rx.combineLatest3(
        menuItemPricesStream,
        apsOrdersStream,
        apsOrderItemsStream,
        (List<MenuItemPrice> menuItems, List<ApsOrder> apsOrders, List<ApsOrderItem> apsOrderItems) {
          List<int> itemIds = menuItems.map((menuItem) => menuItem.itemId).toList();

          return itemIds;
        },
      ).switchMap((List<int> itemIds) {
        return Stream.fromFuture(_supabaseFunctionRepository.getAvailableItems(itemIds: itemIds));
      });
    });
  }

  Stream<List<MenuItemWithDescription>> menuItemsWithDescriptionsStream({
    required int apsId,
  }) {
    final apsStream = _apsRepository.apsStream(apsId: apsId);

    return apsStream.switchMap((apsDescription) {
      final menuId = apsDescription.menuId;

      final menuItemPricesStream = _menuRepository.menuItemPriceStream(menuId: menuId);

      return menuItemPricesStream.switchMap((menuItems) {
        List<int> itemIds = menuItems.map((menuItem) => menuItem.itemId).toList();

        final itemDescriptionsStream = _itemRepository.itemDescriptionsStream(itemIds);

        return Rx.combineLatest2<List<MenuItemPrice>, List<ItemDescription>, List<MenuItemWithDescription>>(
          menuItemPricesStream,
          itemDescriptionsStream,
          (menuItems, itemDescriptions) {
            return menuItems.map((menuItem) {
              final itemDescription = itemDescriptions.firstWhere(
                (description) => description.id == menuItem.itemId,
              );
              return MenuItemWithDescription(
                menuItemPrice: menuItem,
                itemDescription: itemDescription,
              );
            }).toList();
          },
        );
      });
    });
  }

  Future<void> cancelOrder() async {
    // Pobierz aktualne zamówienie ze strumienia
    final order = _orderSubject.valueOrNull;

    if (order != null && order.id != null) {
      try {
        // Aktualizacja statusu zamówienia na "canceled"
        final updateOrderStatus = _orderRepository.updateApsOrder(
          orderId: order.id!,
          data: {
            'status': OrderStatus.canceled.value,
          },
        );

        // Aktualizacja statusu wszystkich pozycji zamówienia na "canceled"
        final updateOrderItemsStatus = _orderRepository.deleteApsOrderItems(
          orderId: order.id!,
        );

        // Poczekaj na zakończenie obu operacji
        await Future.wait([updateOrderStatus, updateOrderItemsStatus]);

        // Zaktualizowanie stanu zamówienia, ustawienie na null
        _orderSubject.add(null);
      } catch (e) {
        print('Error canceling order: $e');
        throw OrderException('Failed to cancel the order');
      }
    }
  }

  Future<void> addItemToOrder(int itemId) async {
    final orderId = _orderSubject.valueOrNull?.id;
    if (orderId != null) {
      final newOrderItem = ApsOrderItem(
        apsId: AppConfig.instance.apsId,
        apsOrderId: orderId,
        itemId: itemId,
        status: ItemStatus.reserved,
      );

      await _orderRepository.createOrderItem(orderItem: newOrderItem);
    }
  }

  Future<void> removeItemFromOrder(int orderItemId) async {
    await _orderRepository.deleteOrderItem(orderItemId: orderItemId);
  }

  Future<void> updatePhoneNumber(String phoneNumber) async {
    final orderId = _orderSubject.valueOrNull?.id;
    if (orderId != null) {
      await _orderRepository.updateApsOrder(
        orderId: orderId,
        data: {
          'client_phone_number': phoneNumber,
        },
      );
    }
  }

  Future<void> updateOrderStatus(OrderStatus newStatus) async {
    final orderId = _orderSubject.valueOrNull?.id;
    if (orderId != null) {
      await _orderRepository.updateApsOrder(
        orderId: orderId,
        data: {
          'status': newStatus.value,
        },
      );
    }
  }

  void finishOrder() {
    _orderSubject.add(null);
  }

  Future<int?> getKDSOrderNumber() async {
    // TODO: moze wyrzucanie exception? i jak nie bedzie zamowienia to ekran strartowy by zaczac od nowa?
    final orderId = _orderSubject.valueOrNull?.id;
    if (orderId != null) {
      final int orderNumber = await _orderRepository.getKdsOrderNumber(apsId: AppConfig.instance.apsId);
      await _orderRepository.updateApsOrder(
        orderId: orderId,
        data: {
          'kds_order_number': orderNumber,
        },
      );
      return orderNumber;
    }
    return null;
  }
}

File: services/order_service.py
from services.supabase_client_manager import SupabaseClientManager

class OrderService:
    def __init__(self):
        self.supabase = SupabaseClientManager.create_local_client()

    def create_order(self, aps_id, origin):
        try:
            response = self.supabase.table('aps_order').insert({
                'aps_id': aps_id,
                'origin': origin,
                'status': 'during_ordering'
            }).execute()
            
            if response.data:
                return response.data[0]['id']
            else:
                return None
        except Exception as e:
            print(f"Error creating order: {e}")
            return None

    def cancel_order(self, order_id):
        try:
            response = self.supabase.table('aps_order').update({'status': 'canceled'}).eq('id', order_id).execute()
            return len(response.data) > 0
        except Exception as e:
            print(f"Error canceling order: {e}")
            return False

    def update_order_items(self, order_id, cart):
        try:
            # Najpierw usuńmy wszystkie istniejące pozycje zamówienia
            self.supabase.table('aps_order_item').delete().eq('aps_order_id', order_id).execute()
            
            # Teraz dodajmy nowe pozycje
            items_to_insert = [
                {'aps_order_id': order_id, 'item_id': int(item_id), 'quantity': quantity, 'status': 'reserved'}
                for item_id, quantity in cart.items() if quantity > 0
            ]
            
            if items_to_insert:
                self.supabase.table('aps_order_item').insert(items_to_insert).execute()
            
            return True
        except Exception as e:
            print(f"Error updating order items: {e}")
            return False

    def get_order_items(self, order_id):
        try:
            response = self.supabase.table('aps_order_item').select('*').eq('aps_order_id', order_id).execute()
            return response.data
        except Exception as e:
            print(f"Error getting order items: {e}")
            return []

    def get_order_total(self, order_id):
        try:
            response = self.supabase.rpc('get_order_total', {'input_order_id': order_id}).execute()
            return response.data
        except Exception as e:
            print(f"Error getting order total: {e}")
            return 0

    def update_order_status(self, order_id, status, kds_order_number=None):
        try:
            update_data = {'status': status}
            if kds_order_number is not None:
                update_data['kds_order_number'] = kds_order_number
            
            response = self.supabase.table('aps_order').update(update_data).eq('id', order_id).execute()
            return len(response.data) > 0
        except Exception as e:
            print(f"Error updating order status: {e}")
            return False

    def get_order_details(self, order_id):
        try:
            response = self.supabase.table('aps_order').select('kds_order_number', 'estimated_time', 'pickup_number').eq('id', order_id).execute()
            if response.data:
                return response.data[0]
            else:
                return None
        except Exception as e:
            print(f"Error getting order details: {e}")
            return None